File: prism-sdk/credentials/shared/src/main/scala/io/iohk/atala/prism/credentials/CredentialBatchId.scala
package io.iohk.atala.prism.credentials

import java.util.UUID

import com.google.protobuf.ByteString
import io.iohk.atala.prism.crypto.MerkleTree.MerkleRoot
import io.iohk.atala.prism.crypto.Sha256Digest
import io.iohk.atala.prism.identity.DIDSuffix
import io.iohk.atala.prism.protos.node_models

class CredentialBatchId private (val id: String) extends AnyVal {
  override def toString: String = id
}

object CredentialBatchId {
  private val CREDENTIAL_BATCH_ID_RE = "^[0-9a-f]{64}$".r

  def fromString(id: String): Option[CredentialBatchId] = {
    if (CREDENTIAL_BATCH_ID_RE.pattern.matcher(id).matches()) Some(new CredentialBatchId(id))
    else None
  }

  def unsafeFromString(string: String): CredentialBatchId = {
    fromString(string).getOrElse(throw new IllegalArgumentException(s"Invalid CredentialBatchId $string"))
  }

  def fromDigest(digest: Sha256Digest): CredentialBatchId = {
    new CredentialBatchId(digest.hexValue)
  }

  def fromBatchData(issuerDIDSuffix: DIDSuffix, merkleRoot: MerkleRoot): CredentialBatchId = {
    val digest = Sha256Digest
      .compute(
        node_models
          .CredentialBatchData()
          .withIssuerDid(issuerDIDSuffix.value)
          .withMerkleRoot(ByteString.copyFrom(merkleRoot.hash.value.toArray))
          .toByteArray
      )

    fromDigest(digest)
  }

  def random(): CredentialBatchId =
    new CredentialBatchId(Sha256.compute(UUID.randomUUID().toString.getBytes()).hexValue)
}

File: prism-kotlin-sdk/credentials/src/commonMain/kotlin/io/iohk/atala/prism/kotlin/credentials/CredentialBatchId.kt
package io.iohk.atala.prism.kotlin.credentials

import com.benasher44.uuid.uuid4
import io.iohk.atala.prism.kotlin.crypto.MerkleRoot
import io.iohk.atala.prism.kotlin.crypto.SHA256Digest
import io.iohk.atala.prism.kotlin.identity.DIDSuffix
import io.iohk.atala.prism.kotlin.protos.CredentialBatchData
import pbandk.ByteArr
import pbandk.encodeToByteArray
import kotlin.js.JsExport
import kotlin.jvm.JvmStatic

@JsExport
data class CredentialBatchId private constructor(val id: String) {

    companion object {
        private val CREDENTIAL_BATCH_ID_RE = Regex("^[0-9a-f]{64}$")

        @JvmStatic
        fun fromString(id: String): CredentialBatchId? =
            if (id.matches(CREDENTIAL_BATCH_ID_RE))
                CredentialBatchId(id)
            else
                null

        @JvmStatic
        fun fromDigest(digest: SHA256Digest): CredentialBatchId =
            CredentialBatchId(digest.hexValue())

        @JvmStatic
        fun fromBatchData(issuerDIDSuffix: DIDSuffix, merkleRoot: MerkleRoot): CredentialBatchId {
            val digest = SHA256Digest
                .compute(
                    CredentialBatchData(
                        issuerDid = issuerDIDSuffix.value,
                        merkleRoot = ByteArr(merkleRoot.hash.value)
                    ).encodeToByteArray()
                )

            return fromDigest(digest)
        }

        @JvmStatic
        fun random(): CredentialBatchId =
            CredentialBatchId(
                SHA256Digest.compute(uuid4().toString().encodeToByteArray()).hexValue()
            )
    }
}
